server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java

```diff
@@ -321,37 +319,37 @@ default void storingMessageWithSameKeyTwiceShouldUpdateMessageAttributes() throw
     @Test
     default void storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult(ExecutorService executorService) throws Exception {
         MailRepository testee = retrieveRepository();
-
+        int nbKeys = 2;
+        int nbIterations = 1000;
+        int threadCount = 20;
         ConcurrentHashMap.KeySetView<String, Boolean> expectedResult = ConcurrentHashMap.newKeySet();
 
-        int nbKeys = 200;
-        Function<Integer, CheckedRunnable> add = (Integer i) -> () -> {
-            String key = computeKey(nbKeys, i);
+        Random random = new Random();
+        ThrowingRunnable add = () -> {
+            String key = computeKey(nbKeys, random.nextInt());
             testee.store(createMail(key));
             expectedResult.add(key);
         };
 
-        Function<Integer, CheckedRunnable> remove = (Integer i) -> () -> {
-            String key = computeKey(nbKeys, i);
+        ThrowingRunnable remove = () -> {
+            String key = computeKey(nbKeys, random.nextInt());
             testee.remove(key);
             expectedResult.remove(key);
         };
 
-        int nbIterations = 1000;
-        Future.sequence(
-            Partition
-                .create(ImmutableMultimap.of(2, add, 6, remove))
-                .generateRandomStream()
-                .zipWithIndex()
-                .map(x -> x._1.apply(x._2))
-                .take(nbIterations)
-                .map(runnable -> io.vavr.concurrent.Future.run(executorService, runnable)))
-            .await();
+        DiscreteDistribution<ThrowingRunnable> distribution = DiscreteDistribution.create(
+            ImmutableList.of(
+                new DistributionEntry<>(add, 2),
+                new DistributionEntry<>(remove, 6)));
```

Comment:
what does mean a probability of 2 or 6? I thought probabilities should be between 0 and 1?

Reply:
Probability is a ratio. A ratio is between 0 and 1. But you can express this ratio with different means . Here 2/8 and 6/8 is perfectly valid.

However, with recent changes I can explain the same thing with 0 <-> 1 values as it supports double.